src/core/thread/mle.cpp

```diff
@@ -890,17 +882,20 @@ void Mle::ApplyMeshLocalPrefix(void)
 #endif
 
 #if OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
-
-    for (Ip6::Netif::UnicastAddress &serviceAloc : mServiceAlocs)
+    for (ServiceAloc &serviceAloc : mServiceAlocs)
     {
-        if (serviceAloc.GetAddress().GetIid().GetLocator() != Mac::kShortAddrInvalid)
+        if (serviceAloc.IsInUse())
         {
             Get<ThreadNetif>().RemoveUnicastAddress(serviceAloc);
-            serviceAloc.GetAddress().SetPrefix(GetMeshLocalPrefix());
+        }
+
+        serviceAloc.ApplyMeshLocalPrefix(GetMeshLocalPrefix());
```

Comment:
Mesh local prefix is always set on all `ServiceAloc` instances in `mServiceAlocs` array independent of whether they are currently in use or not. This way, we just need to `SetAloc16()` on an unused `ServiceAloc` before adding it to `ThreadNetif`.  The IID part of the address is initialized from `ServiceAloc` constructor call.